### PR TITLE
Reflectively load Optional SSL Dependencies

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
Motivation:
Static analyzers complain if there are references to absent classes
in a jar file.  It is possible to remove said files, but for the
SSL handlers in netty doing so only raises more warnings.

Modification:
Jdk(Alpn|Npn) Ssl classes are modified to be more defensive about
loading non present SSL Engines via reflection.

Result:
Less complainy static analysis

This change is a slightly more burdensome change, since compile
time analysis is intentionally broken.  I understand if you might
not want this PR, though since it is useful to my team, I thought
I would pose it.

### CODE REVIEW

File: handler/src/main/java/io/netty/handler/ssl/JdkAlpnApplicationProtocolNegotiator.java [hunk 1]
```diff
@@ -15,14 +15,19 @@
...
 import javax.net.ssl.SSLEngine;
 
+
 /**
  * The {@link JdkApplicationProtocolNegotiator} to use if you need ALPN and are using {@link SslProvider#JDK}.
  */
 public final class JdkAlpnApplicationProtocolNegotiator extends JdkBaseApplicationProtocolNegotiator {
-    private static final boolean AVAILABLE = ConscryptAlpnSslEngine.isAvailable() || JettyAlpnSslEngine.isAvailable();
-    private static final SslEngineWrapperFactory ALPN_WRAPPER = AVAILABLE ? new AlpnWrapper() : new FailureWrapper();
+    private static final SslEngineWrapperFactory ALPN_WRAPPER = new AlpnWrapper();
```
Comment (netkins) [inline handler/src/main/java/io/netty/handler/ssl/JdkAlpnApplicationProtocolNegotiator.java]: MAJOR Avoid instantiation through private constructors from outside of the constructors class. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AAccessorClassGeneration)

**Comment by normanmaurer** [comment]
@trustin @Scottmitch WDYT ?

**Comment by carl-mastrangelo** [comment]
One alternative is to move the reflection directly into the Jetty* and Conscrypt* classes.  They already do reflection, but the amount of change would be much larger that this version.

**Comment by normanmaurer** [comment]
@carl-mastrangelo I kind of like the idea of having all reflection in in one place tho

**Comment by carl-mastrangelo** [comment]
@normanmaurer I investigated if the reflection could be put into the leaf classes.  For Jetty, it would be possible, but would be significantly uglier.  The code would need to resort to using dynamic Proxies in order to create `ALPN.ServerProvider` and `ALPN.ClientProvider` instances.  

For conscrypt, it won't work because it depends on an abstract class `HandshakeListener` which cannot be instantiated reflectively.  From looking around, using javassist could be used to create one, but that may be a bridge too far.

**Comment by Scottmitch** [comment]
We already depend upon properties of the class loaders to delay class loading until first reference for features which are available in newer JDK versions and for our other native libraries. I'm not sure this PR provides much value at the expense of making the class references implicit through reflection. If we get it wrong and load classes which are not available that would be a programing error, and we will fix it, but IIUC for now we are OK.

**Comment by normanmaurer** [comment]
@carl-mastrangelo I agree with @Scottmitch ... WDYT ?

**Comment by carl-mastrangelo** [comment]
I can roll this back.

**Pull request closed without merging**